Fluid interface for building routes in PHP

weaverryan opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | not yet... |

This - along with #15742 - attempts to making adding routes in PHP (via an actual class+method) not only possible, but also useful.

The two classes - `Route` and `RouteCollectionBuilder` are based off of Silex's `Controller` and `ControllerCollection`. The `RouteCollectionBuilder` is basically a `RouteCollection` that's able to import other resources. Here are the goals:

A) Import routes easily

``` php
$routes->import('routing.yml');
```

B) Fluid addition of routes into the collection 

``` php
$routes->add('/admin', 'AppBundle:Admin:index', 'admin_index')
    ->setMethods(['GET']);
```

C) Ability to create routes with auto-generating names

D) Ability to add a "sub-collection" (kind of like an import, without pointing to another file). Included is the ability to set the controller class:

``` php
$blogRoutes = $routes->createBuilder('/blog')
   ->setControllerClass('AppBundle\Controller\BlogController');
$blogRoutes->add('/', 'indexAction');
$blogRoutes->add('/{id}', 'editAction');
$routes->addBuilder($blogRoutes);
```

E) The collection options can be set before or after the routes. With `RouteCollection`, if you set something - e.g. a prefix or a default - and THEN add more routes, those options are not passed to those routes. This is by design, but not ideal for building routes (e.g. in the previous code example, the controllerClass would not be applied using the opposite logic, since it's set before adding the routes).

Thanks!

weaverryan committed:
Adding a new framework-specific Route class

This will join with future commits and a RouteCollectionBuilder

weaverryan committed:
Adding a class to make adding/importing routes easier and more fluid

This is modeled off of Silex's ControllerCollection. The main advantages
over using the raw RouteCollection are:

* Ability to create a Route using a fluid interface (add)
* Ability to import a resource, add that collection and modify it using a fluid interface (import)
* Support for auto-generating route names

weaverryan committed:
Maintaining all the RouteCollection abilities to RouteCollectionBuilder

This allows us to add all of these values AFTER all of the routes have been added,
which is the behavior that makes more sense. Also adding a feature so that you can
mount "collections" or routes at a time, and set the controllerClass on those to
save coding routes from an entire class.

weaverryan committed:
Moving the prefix to the builder, so that it's consistent with other behavior

This is an edge case, but in theory, now if you added more routes to this
same embedded RouteCollectionBuilder, those would also have the prefix.

weaverryan committed:
Adding phpdoc

weaverryan committed:
Using InvalidArgumentException

weaverryan committed:
No change - renaming variable

weaverryan committed:
Adding many more tests, which included a few small bug fixes with values on compile

weaverryan committed:
Removed prefix argument from mount() - and added it instead to createCollection()

The motivation is to avoid confusion: the "prefix" is a property of the embedded
RouteCollectionBuilder itself - and so setting it, mutates that object. With mount(),
it may have appeared that the RouteCollectionBuilder your mounting isn't mutated
with the prefix - but that this value is somehow stored on the parent.

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,393 @@
+    /**
+     * @param Loader $loader
+     */
+    public function __construct(Loader $loader)
```
Like in #15742, this could be `LoaderInterface` instead... which I believe would be ok, since I think we could still "import", just using the longer syntax below (which the user wouldn't need to be bothered with).

stof replied:
IMO, the interface should be used. We should not encourage typehinting against implementations

weaverryan commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,393 @@
+     * @param string $name          The name to give this route
+     * @return Route
+     */
+    public function add($path, $controller, $name = null)
```
We could potentially also have a `get()` - I was trying to minimize the API, but it would allow you to do things like mutate imported routes.

weaverryan committed:
fabbot!

sstok commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,393 @@
+use Symfony\Component\Routing\RouteCollection;
+
+/**
+ * Helps add and import routes into a RouteCollection..
```
duplicate dot ;)

sstok commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,393 @@
+     *
+     * @return RouteCollection
+     */
+    public function flush()
```
This name is a bit strange, most time build or get is used.
You are not really flushing anything.

sstok commented on src/Symfony/Bundle/FrameworkBundle/Routing/Route.php:
```diff
@@ -0,0 +1,101 @@
+    {
+        $methods = implode('_', $this->getMethods()).'_';
+
+        $routeName = $methods.$this->getPath();
```
You may also want to include the hostname requirement, if this is possible.

sstok commented:
:+1: for this feature, really like this :)

stloyd commented:
Hmmmm... http://ocramius.github.io/blog/fluent-interfaces-are-evil/

stof commented:
@stloyd read the blog post you linked. Builders are precisely the place where they are not evil

stloyd commented:
@stof I read that blog post & I checked the proposed code, and you can see that there is builder & new class `Route` which is not real builder (if it would be real builder I would not mention this post)...

weaverryan committed:
Renaming flush() to build()

weaverryan committed:
Not clearing everything on build - unnecessary, and the RouteCollectionBuilder is not meant to be re-used

weaverryan committed:
Fixing phpdoc

weaverryan committed:
Removing the FrameworkBundle Route and the ability to call Route::setName()

Calling Route::setName() (because this value normally lives on the RouteCollection)
made things uglier between normal Route objects and these new Route objects with setName()
methods. To support both in RouteCollectionBuilder, we would have needed to do some
weird things.

Overall, this reduced the scope of the pull request, without too much penalty.

weaverryan committed:
Simplifying by transforming RouteCollection's into RouteCollectionBuilder's

This means that the RouteCollectionBuilder only nests Routes and RouteCollectionBuilders.
It also means that when you import, those routes are added at the top level, whereas before
we added the RouteCollection directly into the builder.

weaverryan committed:
Fixing a bug with knowing which keys should be auto-generated

weaverryan committed:
Minor code improvement to centralize things

weaverryan committed:
Renaming methods for clarity and consistency

weaverryan committed:
fabbot and possible test fixes

weaverryan commented:
Ping @symfony/deciders 

There are some deprecated warnings being triggered by this PR on some PHP version - I haven't tracked that down yet, but it's a non-issue.

stof commented:
@weaverryan IMO, it is not a non issue

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,433 @@
+        $controller = $route->getDefault('_controller');
+
+        // only apply controller class to a (non-empty) string
+        if (!is_string($controller) || !$controller) {
```
I would say ``'' === $controller``. It would be more readable

weaverryan replied:
It's not quite the same thing - I'm also filtering for other things, like array()s

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,433 @@
+        if (false !== strpos($controller, ':')) {
+            return;
+        }
+
```
there is one case not supported here: passing the id of an invokable service

weaverryan replied:
You're absolutely right good catch. But there's no way to support that here unfortunately, and I think that's fine: if you're passing invokable service ids to your routes, you shouldn't set the controller class on your collection. That just doesn't make sense.

In a perfect world, I'd like to be able to detect this and throw an exception, but as this is an extra optional feature, I think if we can make this work for the 99% and allow the 1% to do their work just as easily, that's a win.

weaverryan commented:
@stof without looking into it too far, you don't know why deprecation warnings would be thrown only when "deps" isn't set (deps=high and deps=low looks good)? I can't trigger the same deprecation warnings locally with this code, nor can I see how it's possible that this part of the code is being executed. It is an issue that will need to be resolved before merging, but fundamentally, the PR won't need to use deprecated functionality. I just have to track it down :)

weaverryan committed:
Allowing LoaderInterface instead of Loader

weaverryan committed:
Updating setRequirements to avoid deprecated calls

The problem is that Route::setMethods() ultimately sets the _method requirement for BC. But
then, calling $route->getRequirements() later would return _method, which we'd then set
on the route. This avoids that.

weaverryan commented:
Nevermind - tests should now be green.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     */
+    public function addResource(ResourceInterface $resource)
+    {
+        $this->resources[] = $resource;
```
I think we should check before if the configured loader actually supports the given resource.

weaverryan replied:
This purposefully mirrors `RouteCollection::addResource()`, and I think we should keep it dumb: I could choose to add a "resource" here for the purposes of cache-clearing, even if my loader doesn't support that resource (perhaps *I* know better than the loader, that this resource was in fact used to determine some route information).

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+    }
+
+    /**
+     * Set a controller class that all added embedded should use.
```
embedded routes

weaverryan committed:
phpdoc typo

xabbuh commented:
LGTM :+1:

weaverryan commented:
Status: Reviewed

I'm really hoping this will be considered for 2.8 to make adding PHP routes easier. I've already trimmed it down a lot to make it an easier merge. If I need to, I can remove the `setControllerClass()` functionality, as it looks like the most controversial part.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+    /**
+     * @param LoaderInterface $loader
+     */
+    public function __construct(LoaderInterface $loader)
```
What about making `$loader` optional? If not passed, throw an exception when the user calls `import()`. Everything else would still work.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+    private $methods;
+
+    private $resources = array();
+
```
All blank lines between properties should be removed (I promise to also make more useful comments :)).

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Routing;
```
Why not moving this to the Routing component directly? I don't see any requirements here that would forbid that.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return RouteCollectionBuilder
+     */
+    public function createBuilder($prefix = null)
```
I would move the `$prefix` to `addBuilder` instead:

```php
$blogRoutes = $routes->createBuilder();
$blogRoutes->add('/', 'indexAction');
$blogRoutes->add('/{id}', 'editAction');
$routes->addBuilder('/blog', $blogRoutes);
```

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @param RouteCollectionBuilder $builder
+     */
+    public function addBuilder(RouteCollectionBuilder $builder)
```
`addBuilder()` is really ugly as a name. What about using `mount()` like in Silex?

GromNaN replied:
:+1: `mount` is a very intuitive name.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return RouteCollectionBuilder
+     */
+    public function import($resource, $prefix = null, $type = null)
```
Based on my comments on `createBuilder` and `addBuilder`, I would remove the `$prefix` argument and use composition instead:

```php
$routes->mount('/blog', $routes->import('blog.yml');
```

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return $this
+     */
+    public function addRouteCollection(RouteCollection $collection, $prefix = null)
```
I would remove this one from the public API; the `createBuilder()` one is probably enough. If not, why would I use a builder instead of creating a route collection directly?

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return $this
+     */
+    public function addRoute(Route $route, $name = null)
```
In Symfony, we try to use simple `add/get/set/has` for the main object being manipulated by a collection. So, I would expect `add()` here.

fabpot replied:
Just seen the `add()` method, nvm then

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return $this
+     */
+    public function setPrefix($prefix)
```
I don't think this one is useful at all. If you want a prefix, that's probably a subset of your routes that you've created via `createBuilder` anyway.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+     *
+     * @return $this
+     */
+    public function setControllerClass($controllerClass)
```
I don't like this one at all. I find it confusing and not that useful (given all the conventions we already have to describe a controller).

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+    {
+        $methods = implode('_', $route->getMethods()).'_';
+
+        $routeName = $methods.$route->getPath();
```
The prefix is not yet part of `getPath()` here as it is added later in the process, so we can have collisions here. You should either pass the prefix and add it here manually, or call `generateRouteName()` later in `build()`.

weaverryan replied:
good catch - I've added a test for this along with the fix

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+    public function import($resource, $prefix = null, $type = null)
+    {
+        /** @var RouteCollection $subCollection */
+        $subCollection = $this->resolve($resource, $type)->load($resource, $type);
```
Why not have a private `load()` method that loads the resource; the chain call looks weird to me even more that you need to pass the same arguments to both.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,465 @@
+            throw new FileLoaderLoadException($resource);
+        }
+
+        return $loader;
```
I would simplify this block to:

```php
if (null === $resolver = $this->loader->getResolver()) {
    throw new FileLoaderLoadException($resource);
}

return $resolver->resolve($resource, $type);
```

fabpot commented:
Status: Needs work

weaverryan committed:
Making RouteCollectionBuilder's LoaderInteface optional

Just throw an exception if they try to import

weaverryan committed:
removing extra spaces

weaverryan committed:
moving into the component

weaverryan committed:
Removing the ability to set a prefix on a builder: that only happens when you mount

addBuilder was also renamed to mount

weaverryan committed:
Removing the prefix from import, and making the user actually put that into the parent builder

weaverryan committed:
Fixing a bug with route name collission because the prefix wasn't accounted for

weaverryan committed:
Refactoring into private method

weaverryan commented:
@fabpot I agree with and made all the changes you suggested. As usual, it basically does the same thing as before, but is a bit simpler.

I moved this into the Routing component: the only thing that makes it not *purely* Routing is that it assumes that your controller is on the `_controller` key that's an optional feature of course). But I still think it's best there: could be useful for Drupal peeps, for example.

I haven't squashed to keep the history for now, but it probably needs a good squashing on merge.

Thanks!

weaverryan committed:
fabbot

fabpot commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,366 @@
+class RouteCollectionBuilder
+{
+    /**
+     * A mixture of different objects that hold routes.
```
I think this can be removed, the @var annotation tells us the same.

fabpot commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,366 @@
+     */
+    public function createBuilder()
+    {
+        $builder = new self($this->loader);
```
just return here, no need for a variable.

fabpot commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,366 @@
+     *
+     * @return $this
+     */
+    public function addResource(ResourceInterface $resource)
```
I don't think it should be part of this class, I fail to see when a user of this class would have to call this method.

weaverryan replied:
I've removed it. Let's let someone ask for it and then consider adding it if/when that happens. I can see a possible use-case: you (instead of creating a proper loader) parse through some files/resources manually in PHP and create routes. You then want those resources you parsed through manually to be resources.

fabpot commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,366 @@
+
+        if (false === $loader) {
+            throw new FileLoaderLoadException($resource);
+        }
```
I still prefer my suggestion from a previous version of your code. Here, you artificially use `false` to detect a special condition, which seems very hacky.

weaverryan replied:
I didn't purposefully ignore your suggestion, it's here https://github.com/symfony/symfony/pull/15778#discussion_r40640221 - but I think it's more complicated. The original code comes from the Config component: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Config/Loader/Loader.php#L70

Anyways, I've updated the algorithm to be even more verbose, to make the different situations more clear. I have little doubt there's a slicker way to write it - which should be more obvious now :).

weaverryan committed:
Small tweaks suggested by fabpot, including removal of addResource(), as it probably should not be needed by the end user

weaverryan committed:
stretching out logic into multiple lines: there's some discussion about how to write this

weaverryan committed:
Fabbot

weaverryan commented:
Status: Needs Review

fabpot commented:
:+1:

xabbuh commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,346 @@
+    public function import($resource, $type = null)
+    {
+        /** @var RouteCollection $collection */
+        $collection = $this->load($resource, $type);
```
The `load()` message can throw a `FileLoaderLoadException` which should be documented in the docblock.

weaverryan replied:
Do we add that here, even though it's a second method call that throws the exception? I'm actually not sure what the standard is here.

xabbuh replied:
I'd say we should do that here. The nested method call is a private method and people will never realise they may have to be aware of the exception if we don't document that here.

weaverryan replied:
makes sense to me - I've added it - easy to remove if others don't want it

weaverryan committed:
Adding throws

xabbuh commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,348 @@
+     * @param string      $controller The route's controller
+     * @param string|null $name       The name to give this route
+     *
+     * @return Route
```
This is a bit counter-intuitive as almost all methods return the builder itself. Though probably there is no better solution.

fabpot replied:
I almost wrote the same comment but I think it makes sense for this one to return the route.

Taluu replied:
The name is disturbing, you don not expect a `Route`. I think `add` should do what `addRoute` does (just adding a route to the collection), and then this one, like in the `ContainerBuilder` should be named `register` or something like that ? cf https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L728

weaverryan replied:
I tried to make this method the most easily guessable, since you will call this much more often than `addRoute()`. If we called this `register()`, I think people would be looking for a method like `add()` or `addRoute()`, and might try calling `addRoute()` instead.

Taluu replied:
It also makes sense, as there is no `set` method in the `ContainerBuilder` too (it's directly `setDefinition`)

fabpot replied:
When looking at the code written with this builder, `add()` looks much better and I think it's quite intuitive.

xabbuh commented:
:+1: 

Status: Reviewed

fabpot commented:
:+1:

aitboudad commented:
:+1:

fabpot commented:
Thank you @weaverryan.

Tobion commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    /**
+     * Creates the final ArrayCollection, returns it, and clears everything.
```
ArrayCollection?

WouterJ commented:
These exciting features really deserve some docs before the stable release :wink:

weaverryan commented:
I will *definitely* be taking care of that :)

Tobion commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,348 @@
+    public function add($path, $controller, $name = null)
+    {
+        $route = new Route($path);
+        $route->setDefault('_controller', $controller);
```
The routing component doesn't know anything about a `_controller`. So this seems misplaced. Such assumption can only be made in the framework bundle.

weaverryan replied:
@Tobion you're right that the Routing component doesn't know about `_controller`. But HttpKernel *does* know about this, and I wanted this to be available for things like Drupal. Since there is no bridge between HttpKernel and Routing, I put it here, which purposefully makes this class useable only by people who follow this convention. It's not perfect, but it's the best place. Of course, if you wanted to use this in another system, you could override this method.

WouterJ replied:
If HttpKernel does, why not putting it in HttpKernel instead?

weaverryan replied:
If we put it in HttpKernel, someone will ask why we don't put it in Routing :)

WouterJ replied:
And then we have a clear answer. HttpKernel contains classes that add just a little "HttpKernel convention sauce" on top of classes provided by other components (DI Extension, File locator, ...)

weaverryan replied:
Good point. My counterpoint is that it makes sense to use this without HttpKernel - maybe you have a system where _controller is meaningful or you override this method. That should be possible without requiring http-kernel, which this class doesn't actually depend on.

So there's no perfect spot, but I think this is best

Tobion commented on src/Symfony/Component/Routing/RouteCollectionBuilder.php:
```diff
@@ -0,0 +1,348 @@
+     *
+     * @return string
+     */
+    private function generateRouteName(Route $route)
```
this method seems very vulnerable to collisions. It only uses the path and methods? What about the scheme, host, condition etc? So two homepages routes under different domains, would have the same name and thus overwrite each other?

weaverryan replied:
It's true, but we can probably not ever consider *all* the conditions. So, the question is do we try? Or do we make this work for most cases, and really, you *should* be naming your routes.

The truth is that I based this off of what SensioFrameworkExtraBundle and Silex does. I don't remember if the logic is exactly the same here, but those are also both vulnerable to collision.
